lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,462 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
-        public void OnNext(IActiveContext activeContext)
+        /// <param name="allocatedEvaluator"></param>
+        private void SubmitContextAndService(IAllocatedEvaluator allocatedEvaluator)
         {
-            Logger.Log(Level.Verbose, string.Format("Received Active Context {0}", activeContext.Id));
-
-            if (_serviceAndContextConfigurationProvider.IsMasterEvaluatorId(activeContext.EvaluatorId))
+            ContextAndServiceConfiguration configs;
+            if (_evaluatorManager.IsEvaluatorForMaster(allocatedEvaluator))
             {
-                Logger.Log(Level.Verbose, "Submitting master task");
-                _commGroup.AddTask(IMRUConstants.UpdateTaskName);
-                _groupCommTaskStarter.QueueTask(GetUpdateTaskConfiguration(), activeContext);
-                RequestMapEvaluators(_totalMappers);
+                configs =
+                    _serviceAndContextConfigurationProvider
+                        .GetContextConfigurationForMasterEvaluatorById(
+                            allocatedEvaluator.Id);
             }
             else
             {
-                Logger.Log(Level.Verbose, "Submitting map task");
-                _serviceAndContextConfigurationProvider.RecordActiveContextPerEvaluatorId(activeContext.EvaluatorId);
-                string taskId = GetTaskIdByEvaluatorId(activeContext.EvaluatorId);
-                _commGroup.AddTask(taskId);
-                _groupCommTaskStarter.QueueTask(GetMapTaskConfiguration(activeContext, taskId), activeContext);
-                Interlocked.Increment(ref _numberOfReadyTasks);
-                Logger.Log(Level.Verbose, string.Format("{0} Tasks are ready for submission", _numberOfReadyTasks));
+                configs = _serviceAndContextConfigurationProvider
+                    .GetDataLoadingConfigurationForEvaluatorById(
+                        allocatedEvaluator.Id);
             }
+            allocatedEvaluator.SubmitContextAndService(configs.Context, configs.Service);
         }
+        #endregion IAllocatedEvaluator
 
+        #region IActiveContext
         /// <summary>
-        /// Specifies what to do when the task is completed
-        /// In this case just disposes off the task
+        /// IActiveContext handler. It will take the following actions based on the system state:
+        /// Case WaitingForEvaluator:
+        ///    Adds Active Context to Active Context Manager
+        /// Case Fail:
+        ///    Closes the ActiveContext
+        /// Other cases - not expected
         /// </summary>
-        /// <param name="completedTask">The link to the completed task</param>
-        public void OnNext(ICompletedTask completedTask)
+        /// <param name="activeContext"></param>
+        public void OnNext(IActiveContext activeContext)
         {
-            lock (_completedTasks)
+            Logger.Log(Level.Verbose, "Received Active Context {0}, systemState {1}.", activeContext.Id, _systemState.CurrentState);
```

Comment:
please check other methods as well, logger statement is often outside of lock

---
In reply to: [67060660](https://github.com/apache/reef/pull/1033#discussion_r67060660) [](ancestors = 67060660)